bitrisescript/src/bitrisescript/bitrise.py

```diff
@@ -0,0 +1,218 @@
+        data = response["data"]
+
+        if data["finished_at"]:
+            log.info(f"Build {build_slug} is finished")
+            break
+
+        log.debug(f"Build {build_slug} is still running, waiting {poll_interval}s...")
+        await asyncio.sleep(poll_interval)
+
+    if data["status_text"] != "success":
+        if data["status_text"] == "error":
+            raise Exception(f"Job {build_slug} errored! Got: {response}")
+        if data["status_text"] == "aborted":
+            raise Exception(f"Job {build_slug} was aborted. Got: {response}")
+        raise Exception(f"Job {build_slug} is finished but not successful. Got: {response}")
+
+
+async def download_artifacts(build_slug: str, artifacts_dir: str) -> None:
+    """Download all artifacts from the given Bitrise build.
+
+    Args:
+        build_slug (str): The Bitrise build to download artifacts from.
+        artifacts_dir (str): Directory to save artifacts to.
+    """
+    client = BitriseClient()
+    endpoint = f"/builds/{build_slug}/artifacts"
+    response = await client.do_request(endpoint)
+
+    artifacts_metadata = {metadata["slug"]: metadata["title"] for metadata in response["data"]}
+
+    for artifact_slug, title in artifacts_metadata.items():
+        endpoint = f"/builds/{build_slug}/artifacts/{artifact_slug}"
+        data = (await client.do_request(endpoint))["data"]
+
+        download_url = data["expiring_download_url"]
+        await download_file(download_url, os.path.join(artifacts_dir, title))
+
+
+async def download_log(build_slug: str, artifacts_dir: str, poll_interval: int = 60) -> None:
+    """Download the log from the given Bitrise build.
+
+    Args:
+        build_slug (str): The Bitrise build to download the log from.
+        artifacts_dir (str): Directory to save artifacts to.
+        poll_interval (int): Time in seconds to poll for completion of log (default: 60).
+    """
+    client = BitriseClient()
+    endpoint = f"/builds/{build_slug}/log"
+
+    while True:
+        response = await client.do_request(endpoint)
+        if response["is_archived"] is True:
+            log.info("Log is now ready")
+            break
+        else:
+            log.info("Log is still running. Waiting another minute...")
+            await asyncio.sleep(poll_interval)
+
+    download_url = response["expiring_raw_log_url"]
+    if download_url:
+        await download_file(download_url, os.path.join(artifacts_dir, "bitrise.log"))
+    else:
+        log.error("Bitrise has no log to offer for job {0}. Please check https://app.bitrise.io/build/{0}".format(build_slug))
+
+
+async def download_file(download_url: str, file_destination: str, chunk_size: int = 512) -> None:
+    """Download a file.
+
+    Args:
+        download_url (str): Url to download.
+        file_destination (str): Path to save the file.
+        chunk_size (int): Size in bytes to stream to disk at a time (default: 512).
+    """
+    async with RetryClient() as s3_client:
```

Comment:
I'm not sure why this variable is called `s3`

Reply:
This was copy/pasted from the code in `firefox-ios`. I'm guessing it was because Bitrise is putting its artifacts in S3 storage.. But I agree, that's not a detail we should care about to any degree. I'll update it.